File: sdks/java/src/main/java/do_/mongo/MongoDatabase.java
package do_.mongo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * MongoDB Database - provides access to collections and database operations.
 *
 * <pre>{@code
 * MongoDatabase db = client.getDatabase("myapp");
 *
 * // Get a collection
 * MongoCollection<Document> users = db.getCollection("users");
 *
 * // List collections
 * for (String name : db.listCollectionNames()) {
 *     System.out.println(name);
 * }
 *
 * // Create a collection
 * db.createCollection("logs");
 *
 * // Drop the database
 * db.drop();
 * }</pre>
 */
public class MongoDatabase {

    private final RpcTransport transport;
    private final String name;
    private final Map<String, MongoCollection<?>> collections;

    /**
     * Creates a new MongoDatabase.
     *
     * @param transport the RPC transport
     * @param name      the database name
     */
    MongoDatabase(RpcTransport transport, String name) {
        this.transport = transport;
        this.name = name;
        this.collections = new HashMap<>();
    }

    /**
     * Gets the database name.
     *
     * @return the database name
     */
    public String getName() {
        return name;
    }

    // ============================================================================
    // Collection Access
    // ============================================================================

    /**
     * Gets a collection with Document type.
     *
     * @param collectionName the collection name
     * @return the collection
     */
    public MongoCollection<Document> getCollection(String collectionName) {
        return getCollection(collectionName, Document.class);
    }

    /**
     * Gets a collection with a specific document type.
     *
     * @param collectionName the collection name
     * @param documentClass  the document class
     * @return the collection
     */
    @SuppressWarnings("unchecked")
    public <T> MongoCollection<T> getCollection(String collectionName, Class<T> documentClass) {
        String key = collectionName + ":" + documentClass.getName();
        return (MongoCollection<T>) collections.computeIfAbsent(key,
                k -> new MongoCollection<>(transport, name, collectionName, documentClass));
    }

    // ============================================================================
    // Collection Management
    // ============================================================================

    /**
     * Creates a new collection.
     *
     * @param collectionName the collection name
     */
    public void createCollection(String collectionName) {
        createCollection(collectionName, new Document());
    }

    /**
     * Creates a new collection with options.
     *
     * @param collectionName the collection name
     * @param options        the collection options
     */
    public void createCollection(String collectionName, Document options) {
        transport.call("createCollection", name, collectionName, options);
    }

    /**
     * Creates a new collection asynchronously.
     *
     * @param collectionName the collection name
     * @return a CompletableFuture
     */
    public CompletableFuture<Void> createCollectionAsync(String collectionName) {
        return transport.callAsync("createCollection", name, collectionName, new Document())
                .thenApply(r -> null);
    }

    /**
     * Lists collection names.
     *
     * @return a list of collection names
     */
    @SuppressWarnings("unchecked")
    public List<String> listCollectionNames() {
        Object result = transport.call("listCollections", name, new Document());
        List<String> names = new ArrayList<>();
        if (result instanceof List) {
            for (Object item : (List<?>) result) {
                if (item instanceof Document) {
                    names.add(((Document) item).getString("name"));
                } else if (item instanceof Map) {
                    names.add((String) ((Map<?, ?>) item).get("name"));
                }
            }
        }
        return names;
    }

    /**
     * Lists collections with full information.
     *
     * @return a list of collection info documents
     */
    @SuppressWarnings("unchecked")
    public List<Document> listCollections() {
        Object result = transport.call("listCollections", name, new Document());
        List<Document> collections = new ArrayList<>();
        if (result instanceof List) {
            for (Object item : (List<?>) result) {
                if (item instanceof Document) {
                    collections.add((Document) item);
                } else if (item instanceof Map) {
                    collections.add(new Document((Map<String, Object>) item));
                }
            }
        }
        return collections;
    }

    /**
     * Renames a collection.
     *
     * @param oldName the current collection name
     * @param newName the new collection name
     */
    public void renameCollection(String oldName, String newName) {
        renameCollection(oldName, newName, false);
    }

    /**
     * Renames a collection.
     *
     * @param oldName    the current collection name
     * @param newName    the new collection name
     * @param dropTarget whether to drop the target collection if it exists
     */
    public void renameCollection(String oldName, String newName, boolean dropTarget) {
        Document options = new Document("dropTarget", dropTarget);
        transport.call("renameCollection", name, oldName, newName, options);
        // Update cache
        collections.remove(oldName + ":do_.mongo.Document");
    }

    // ============================================================================
    // Database Operations
    // ============================================================================

    /**
     * Drops the database.
     */
    public void drop() {
        transport.call("dropDatabase", name);
        collections.clear();
    }

    /**
     * Drops the database asynchronously.
     *
     * @return a CompletableFuture
     */
    public CompletableFuture<Void> dropAsync() {
        return transport.callAsync("dropDatabase", name)
                .thenApply(r -> {
                    collections.clear();
                    return null;
                });
    }

    /**
     * Gets database statistics.
     *
     * @return the database stats
     */
    @SuppressWarnings("unchecked")
    public Document getStats() {
        Object result = transport.call("runCommand", name, new Document("dbStats", 1));
        if (result instanceof Document) {
            return (Document) result;
        } else if (result instanceof Map) {
            return new Document((Map<String, Object>) result);
        }
        return new Document();
    }

    /**
     * Runs a database command.
     *
     * @param command the command document
     * @return the command result
     */
    @SuppressWarnings("unchecked")
    public Document runCommand(Document command) {
        Object result = transport.call("runCommand", name, command);
        if (result instanceof Document) {
            return (Document) result;
        } else if (result instanceof Map) {
            return new Document((Map<String, Object>) result);
        }
        return new Document();
    }

    /**
     * Runs a database command asynchronously.
     *
     * @param command the command document
     * @return a CompletableFuture with the result
     */
    @SuppressWarnings("unchecked")
    public CompletableFuture<Document> runCommandAsync(Document command) {
        return transport.callAsync("runCommand", name, command)
                .thenApply(result -> {
                    if (result instanceof Document) {
                        return (Document) result;
                    } else if (result instanceof Map) {
                        return new Document((Map<String, Object>) result);
                    }
                    return new Document();
                });
    }
}

File: sdks/kotlin/src/main/kotlin/do_/mongo/MongoDatabase.kt
package do_.mongo

/**
 * MongoDB Database - provides access to collections.
 *
 * Example usage:
 * ```kotlin
 * val db = client.getDatabase("myapp")
 *
 * // Get a collection
 * val users = db.getCollection<Document>("users")
 *
 * // List collections
 * val collections = db.listCollectionNames()
 *
 * // Create a collection
 * db.createCollection("logs")
 * ```
 */
class MongoDatabase(
    internal val transport: RpcTransport,
    internal val dbName: String
) {
    /**
     * Gets the database name.
     */
    val name: String get() = dbName

    /**
     * Gets a collection by name.
     */
    fun getCollection(name: String): MongoCollection<Document> =
        MongoCollection(transport, dbName, name, Document::class.java)

    /**
     * Gets a typed collection.
     */
    fun <T : Any> getCollection(name: String, documentClass: Class<T>): MongoCollection<T> =
        MongoCollection(transport, dbName, name, documentClass)

    /**
     * Lists all collection names.
     */
    @Suppress("UNCHECKED_CAST")
    fun listCollectionNames(): List<String> {
        val result = transport.call("listCollections", dbName)
        return when (result) {
            is List<*> -> result.mapNotNull { item ->
                when (item) {
                    is Document -> item.getString("name")
                    is Map<*, *> -> item["name"]?.toString()
                    is String -> item
                    else -> null
                }
            }
            else -> emptyList()
        }
    }

    /**
     * Lists all collection names asynchronously.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun listCollectionNamesAsync(): List<String> {
        val result = transport.callAsync("listCollections", dbName)
        return when (result) {
            is List<*> -> result.mapNotNull { item ->
                when (item) {
                    is Document -> item.getString("name")
                    is Map<*, *> -> item["name"]?.toString()
                    is String -> item
                    else -> null
                }
            }
            else -> emptyList()
        }
    }

    /**
     * Lists all collections with full info.
     */
    @Suppress("UNCHECKED_CAST")
    fun listCollections(): List<Document> {
        val result = transport.call("listCollections", dbName)
        return when (result) {
            is List<*> -> result.mapNotNull { item ->
                when (item) {
                    is Document -> item
                    is Map<*, *> -> Document(item as Map<String, Any?>)
                    else -> null
                }
            }
            else -> emptyList()
        }
    }

    /**
     * Creates a new collection.
     */
    fun createCollection(name: String, options: Document = Document()) {
        transport.call("createCollection", dbName, name, options)
    }

    /**
     * Creates a new collection asynchronously.
     */
    suspend fun createCollectionAsync(name: String, options: Document = Document()) {
        transport.callAsync("createCollection", dbName, name, options)
    }

    /**
     * Drops the database.
     */
    fun drop() {
        transport.call("dropDatabase", dbName)
    }

    /**
     * Drops the database asynchronously.
     */
    suspend fun dropAsync() {
        transport.callAsync("dropDatabase", dbName)
    }

    /**
     * Runs a command on the database.
     */
    @Suppress("UNCHECKED_CAST")
    fun runCommand(command: Document): Document {
        val result = transport.call("runCommand", dbName, command)
        return when (result) {
            is Document -> result
            is Map<*, *> -> Document(result as Map<String, Any?>)
            else -> Document()
        }
    }

    /**
     * Runs a command on the database asynchronously.
     */
    @Suppress("UNCHECKED_CAST")
    suspend fun runCommandAsync(command: Document): Document {
        val result = transport.callAsync("runCommand", dbName, command)
        return when (result) {
            is Document -> result
            is Map<*, *> -> Document(result as Map<String, Any?>)
            else -> Document()
        }
    }

    override fun toString(): String = "MongoDatabase($dbName)"
}
